Fix 'path' argument. Add more visitor tests

IvanGoncharov opened the pull request:
As discussed in #1145 I started to work on tests for `visit` function.

IvanGoncharov commented:
@leebyron I added the simple test for `path` parameter and discover that `path` value differs between `enter` and `leave` callbacks. On `leave` path is always one element shorter, for example if you was called on enter to `OperationDefinition` node `path` value would be `['definitions', 0]` but on leave you will get `['definitions']`.

Is it intentional behaviour or a bug?

mjmahone commented on src/language/__tests__/visitor-test.js:
```diff
@@ -17,6 +17,35 @@ import { testSchema } from '../../validation/__tests__/harness';
+      },
+    });
+
+    console.log(JSON.stringify(visited, null, 2));
```
Probably don't want to keep this console.log in the test

IvanGoncharov replied:
@mjmahone Consequences of committing at 3 AM 😄 
Removed.

mjmahone commented on src/language/__tests__/visitor-test.js:
```diff
@@ -17,6 +17,35 @@ import { testSchema } from '../../validation/__tests__/harness';
 import { getNamedType, isCompositeType } from '../../type';
 
 describe('Visitor', () => {
+  it('check path argument', () => {
```
nit: 'validates path argument' or 'checks path argument'

IvanGoncharov replied:
@mjmahone Thanks 👍 Fixed now.

IvanGoncharov commented:
I fixed this with `path` by a very small fix that's easy to verify.

As for testing visitor callback arguments, I decided it's too much to add it hardcoded snapshots inside tests. So I added `checkVisitorFnArgs` utility function that checks types and interdependencies of all callback arguments and calls it from every callback inside `visitor-tests.js`.

One thing to note is the current implementation of `visit` doesn't maintain consistency of arguments after AST is edited. For example, if you edit some node on `enter` you would get updated node on `leave` but `parent` and all ancestors will remain not updated. To workaround this problem I added `isEdited` argument that disables some of the checks inside `checkVisitorFnArgs` function.

@mjmahone @leebyron Can you please review this PR?

leebyron commented:
Nice work, @IvanGoncharov
